File: Grafos/DFS/DAA027.cpp
#include <iostream>

using namespace std;

#define MAX 55
#define RED 0
#define GREEN 1

int n;              // Número de nós
bool m[MAX][MAX];   // Matriz de adjacências
bool visited[MAX];  // Visitados
int color[MAX];     // Cores dos nós

void reset() {
    for(int i = 0; i < MAX; i++) {
        visited[i] = false;
        color[i] = -1;
        for(int j = 0; j < MAX; j++)
            m[i][j] = false;
    }
}

bool dfs(int v, int c) {
    if(visited[v]) {
        if(color[v] == c) return true;
        return false;
    } else {
        visited[v] = true;
        color[v] = c;
        (c == RED) ? c = GREEN : c = RED;
        for(int i = 1; i <= n; i++)
            if(m[v][i] && !dfs(i,c)) return false;
        return true;
    }
}

int main() {
    int cases;  // Número de casos a considerar
    scanf("%d", &cases);
    for(int c = 0; c < cases; c++) {
        int e, a, b;    // Número de arestas, ligações
        reset();
        scanf("%d %d", &n, &e);
        for(int i = 0; i < e; i++) {
            scanf("%d %d", &a, &b);
            m[a][b] = m[b][a] = true;   // Grafo não dirigido
        }

        if(dfs(1,RED))
            printf("sim\n");
        else 
            printf("nao\n");
    }
    return 0;
}

File: Grafos/DFS/DAA027.java
import java.util.*;

public class DAA027 {
    static int n;               // Número de nós
    static boolean[][] m;       // Matriz de adjacências
    static boolean[] visited;   // Visitados
    static int[] color;         // Cores dos nós
    static final int RED = 0;
    static final int GREEN = 1;

    static boolean dfs(int v, int c) {
        if(visited[v]) {
            if(color[v] == c) 
                return true;
            else 
                return false;
        }
        visited[v] = true;
        color[v] = c;
        c = (c == RED) ? GREEN : RED;
        for(int i = 1; i <= n; i++) 
            if(m[v][i] && !dfs(i,c))
                return false;
        return true;
    }

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int cases = in.nextInt();
        for(int c = 0; c < cases; c++) {
            n = in.nextInt();
            int e = in.nextInt();
            
            m = new boolean[n+1][n+1];
            visited = new boolean[n+1];
            color = new int[n+1];
            
            for(int i = 0; i < color.length; i++) color[i] = -1;

            for(int i = 0; i < e; i++) {
                int a = in.nextInt();
                int b = in.nextInt();
                m[a][b] = m[b][a] = true;   // Grafo não dirigido
            }

            if(dfs(1,RED))
                System.out.println("sim");
            else
                System.out.println("nao");
        }
        in.close();
    }
}
